### PR TITLE
Support returning None from torch.cond() true_fn and false_fn

### LINKED ISSUE #181891: torch.cond()'s true_fn and false_fn are not allowed to return None
### 🐛 Describe the bug

torch.cond() currently fails if its true_fn and false_fn return None.

I realized this when I wrote the following code intended to implement loss scaling for mixed precision training inside a cuda graph:

```
import torch
from torch.utils.data import DataLoader, Dataset
from torch._higher_order_ops.cudagraph_conditional_nodes import (
    ControlFlowOpWarmupDispatchMode,
    CUDAGraphCaptureControlFlowOpDispatchMode,
)

class RandomDataset(Dataset):
    def __init__(self, n_samples=256, in_dim=16, out_dim=4):
        self.n_samples = n_samples
        self.in_dim = in_dim
        self.out_dim = out_dim

    def __len__(self):
        return self.n_samples

    def __getitem__(self, _):
        return torch.randn(self.in_dim), torch.randn(self.out_dim)

model = torch.nn.Linear(16, 4).cuda()
criterion = torch.nn.MSELoss()
data_loader = DataLoader(RandomDataset(), batch_size=32, shuffle=False,
                         pin_memory=True, num_workers=1)


lr = torch.tensor(0.1, device="cuda")
loss_scale = torch.tensor(2.0**8, device="cuda")

def train(X, y):
    with torch.autograd.grad_mode.set_multithreading_enabled(False):
        global loss_scale
        model.zero_grad(set_to_none=True)
        y_pred = model(X)
        loss = criterion(y_pred, y)
        (loss * loss_scale).backward()
        all_gradients_finite = torch.stack([
            torch.all(param.grad.isfinite()) for param in model.parameters() if param.grad is not None
        ]).all()
        with torch.no_grad():
            def true_fn():
                global lr, loss_scale
                lr.mul_(0.99999)
                inv_scale = loss_scale.reciprocal()
                for param in model.parameters():
                    if param.grad is not None:
                        param.sub_(lr * param.grad * inv_scale)
                loss_scale.mul_(2.0)

            def false_fn():
                loss_scale.mul_(0.5)

            _ = torch.cond(all_gradients_finite, true_fn, false_fn)
        return loss

for i, (X, y) in enumerate(data_loader):
    if i == 0:
        with ControlFlowOpWarmupDispatchMode():
            loss = train(X.cuda(), y.cuda())
        print(loss.item())
        del loss
        continue
    if i == 1:
        X_static = torch.empty_like(X, device="cuda")
        y_static = torch.empty_like(y, device="cuda")
        g = torch.cuda.CUDAGraph()
        with torch.cuda.graph(g, capture_error_mode="thread_local"), CUDAGraphCaptureControlFlowOpDispatchMode():
            loss_static = train(X_static, y_static)
    if i == 2:
        torch.cuda.cudart().cudaProfilerStart()
    X_static.copy_(X, non_blocking=True)
    y_static.copy_(y, non_blocking=True)
    g.replay()
    print(loss_static.item())
```

The current restriction is a reasonable oversight, since input mutations in torch.cond() were not supported until https://github.com/pytorch/pytorch/pull/172836. Without input mutations, it doesn't make any sense to return nothing. However, things have changed now. You can see clearly that, for gradient nan checks like above, there is no obvious value to return.

I will upload a fix for this shortly. FYI @ydwu4 @kshitij12345 @kiya00 It is likely that torch.while_loop() has a similar issue, though I haven't tested that yet.

### Versions

top of tree

cc @mcarilli @ezyang @eellison @penguinwu @BoyuanFeng @chauhang @ydwu4 @bdhirsh @bobrenjc93 @aorenste

### PR DESCRIPTION
**Pull request by galv**
This allows us to run conditional code that only mutates inputs or captured variables, but does not return anything.

While making this change, I realized as well that it should be disallowed for true_fn and false_fn to return different non-tensor outputs in the same output slot when using torch.cond() during cuda graph capture. This is banned because a different non-tensor output could cause the cuda graph to be different depending upon which path was taken. I now raise an explicit error alerting the user. Previously, you would simply get a runtime crash if you tried to return a non-tensor output at all, even if it was the same for both true_fn and false_fn.

I use either claude-code or codex to help make these edits. Forgot which though.

Fixes #181891

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @chauhang @amjames @Lucaskabela @jataylo @azahed98 @ydwu4

### CODE REVIEW

File: test/functorch/test_control_flow.py [hunk 1]
```diff
@@ -5665,6 +5665,70 @@ def f(x, p0, p1, p2):
             _check_compile_cudagraph_backend(self, f, args)
             _check_compile_many_backends_with_cudagraph(self, f, args)
 
+    @unittest.skipIf(
+        not TEST_CUDA_GRAPH_CONDITIONAL_NODES,
+        "CUDA 12.4 or greater is required for CUDA Graphs with conditional nodes",
+    )
+    def test_cond_traced_no_outputs_cudagraphs(self):
```
Comment (ydwu4) [inline test/functorch/test_control_flow.py:5673]: Can have a test that runs torch.compile(backend=AotEagerCaptureGraph) and assertExpectedInline the normalize_gm(graph)? Curious if aot_autograd works out of box with the change.
 
Another test worth adding is an inductor test in test/inductor/test_control_flow.py.

**Review by ydwu4 — COMMENTED** [review]
Need some more testing to confirm the correctness.

File: test/functorch/test_control_flow.py (hunk 1 above)
Comment (galv) [reply test/functorch/test_control_flow.py:5673]: Unfortunately, inductor support is missing for torch.cond() + cuda graph right now. It's not a hard one to add, but it hasn't come up yet. I also wonder about aot_autograd. I would expect no, but it's my fault that I did not think it through before hand. I will do it!

File: torch/_higher_order_ops/cudagraph_conditional_nodes.py [hunk 2]
```diff
@@ -126,5 +126,15 @@ def if_else_node(pred: torch.Tensor, true_fn, false_fn, operands):
...
+                        not (
+                            isinstance(if_out, torch.Tensor)
+                            and isinstance(else_out, torch.Tensor)
+                        )
+                        and if_out != else_out
+                    ):
+                        raise ValueError(
+                            f"For cudagraph captured torch.cond(), the returned constants of both branches must be equal. Got true_fn tried to return {if_out} and while false_fn tried to return {else_out}"
+                        )
                     if_out.copy_(else_out)
```
Comment (kshitij12345) [inline torch/_higher_order_ops/cudagraph_conditional_nodes.py]: `if_out.copy_(else_out)` should be guarded to check that if_out and else_out are both Tensors. In case `if_out` or `else_out` are non-Tensor values then `.copy_` would error in that case.

**Comment by jansel** [comment]
@claude explain those CI failures, are they related?

**Comment by galv** [comment]
@claude why is the @skipIfTorchDynamo decorator not applied to the test test_while_loop_autograd_simple in test/functorch/test_control_flow.py? That uses torch.compile() internally as well.

**Comment by galv** [comment]
@ydwu4 once CI passes, this should be good to merge. Can you please approve if you have no more concerns?

**Comment by zou3519** [comment]
I'm deferring to @ydwu4

**Comment by jansel** [comment]
Will we DCE the torch.cond call if the output is unused?  How is mutation/functionalization handled for cond?
